deploy/ray-operator/controllers/raycluster_controller.go

```diff
@@ -0,0 +1,238 @@
+		} else {
+			replicas = append(replicas, pod)
+		}
+	}
+
+	// create service for head
+	if needServicePodMap.Cardinality() > 0 {
+		for elem := range needServicePodMap.Iterator().C {
+			podName := elem.(string)
+			svcConf := common.DefaultServiceConfig(*instance, podName)
+			rayPodSvc := common.ServiceForPod(svcConf)
+			if errSvc := r.Create(context.TODO(), rayPodSvc); errSvc != nil {
+				if errors.IsAlreadyExists(errSvc) {
+					log.Info("Pod service already exist,no need to create")
+				} else {
+					log.Error(errSvc, "Pod Service create error!", "Pod.Service.Error", errSvc)
+					return reconcile.Result{}, errSvc
+				}
+			} else {
+				log.Info("Pod Service create anew successfully", "podName", rayPodSvc.Name)
+			}
+		}
+	}
+
+	// check pod and create one by one if not exist
+	for i, replica := range replicas {
+		// create pod if not exist
+		if !utils.IsCreated(&replica) {
+			log.Info("Creating pod", "index", i, "create pod", replica.Name)
+			if err := r.Create(context.TODO(), &replica); err != nil {
+				return reconcile.Result{}, err
+			}
+			// pod created, no more work possible for this round
+			continue
+		}
+	}
+
+	// delete pods to desired state
+	if difference.Cardinality() > 0 {
+		log.Info("difference", "pods", difference)
+		for _, runtimePod := range runtimePods.Items {
+			if difference.Contains(runtimePod.Name) {
+				log.Info("Deleting pod", "namespace", runtimePod.Namespace, "name", runtimePod.Name)
+				if err := r.Delete(context.TODO(), &runtimePod); err != nil {
+					return reconcile.Result{}, err
+				}
+				if strings.EqualFold(runtimePod.Labels[common.ClusterPodType], common.Head) {
+					svcConf := common.DefaultServiceConfig(*instance, runtimePod.Name)
+					raySvcHead := common.ServiceForPod(svcConf)
+					log.Info("delete head service", "headName", runtimePod.Name)
+					if err := r.Delete(context.TODO(), raySvcHead); err != nil {
+						return reconcile.Result{}, err
+					}
+				}
+			}
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// Build cluster instance pods.
+func (r *RayClusterReconciler) buildPods(instance *rayiov1alpha1.RayCluster) []corev1.Pod {
+	var pods []corev1.Pod
+	if instance.Spec.Extensions != nil && len(instance.Spec.Extensions) > 0 {
+		for _, extension := range instance.Spec.Extensions {
+			var i int32 = 0
+			for i = 0; i < *extension.Replicas; i++ {
+				podType := fmt.Sprintf("%v", extension.Type)
+				podName := instance.Name + common.DashSymbol + extension.GroupName + common.DashSymbol + podType + common.DashSymbol + utils.FormatInt32(i)
+				podConf := common.DefaultPodConfig(instance, podType, podName)
+				podConf.Extension = extension
+				pod := common.BuildPod(podConf)
+				pods = append(pods, *pod)
+			}
+		}
+	} else {
+		log.Info("RayCluster extensions are nil or empty")
+	}
+
+	return pods
+}
+
+func (r *RayClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&rayiov1alpha1.RayCluster{}).
```

Comment:
I think we should add watch here, not in https://github.com/ray-project/ray/pull/6501/files#diff-3930e56b5281ab511234564aa92bdba6R43 if you are using kubebuilder v2 scaffolding